tests: kernel: Add test to validate k_stack_cleanup and k_stack_alloc…

ajaykish opened the pull request:
This test is intended to validate k_stack_cleanup() and k_stack_alloc_init(),
when CONFIG_USERSPACE is not defined .
Passed stack pointer in k_stack_alloc_init() is statically defined.

Signed-off-by: Ajay Kishore <ajay.kishore@intel.com>

andrewboie commented on tests/kernel/stack/stack_api/src/test_stack_contexts.c:
```diff
@@ -6,12 +6,13 @@
 #include <ztest.h>
 #include <irq_offload.h>
-
+#include <syscall_handler.h>
```
Why are you pulling this header?

ajaykish replied:
@andrewboie  Removed.

andrewboie commented on tests/kernel/stack/stack_api/src/test_stack_contexts.c:
```diff
@@ -6,12 +6,13 @@
 
 /**TESTPOINT: init via K_STACK_DEFINE*/
 K_STACK_DEFINE(kstack, STACK_LEN);
+K_STACK_DEFINE(kstack1, STACK_LEN);
```
Please provide a more descriptive name

ajaykish replied:
@andrewboie changed the name kstack1 -> kstack_test_alloc

andrewboie commented on tests/kernel/stack/stack_api/src/main.c:
```diff
@@ -51,6 +52,7 @@ void test_main(void)
 			 ztest_unit_test(test_stack_pop_fail),
-			 ztest_user_unit_test(test_stack_user_pop_fail));
+			 ztest_user_unit_test(test_stack_user_pop_fail),
+			ztest_unit_test(test_stack_alloc_thread2thread));
```
are you using spaces instead of tabs? this should line up

andrewboie commented on tests/kernel/stack/stack_api/prj.conf:
```diff
@@ -2,3 +2,5 @@ CONFIG_ZTEST=y
 CONFIG_USERSPACE=y
 CONFIG_DYNAMIC_OBJECTS=y
+CONFIG_OBJECT_TRACING=y
+CONFIG_HEAP_MEM_POOL_SIZE=128
```
Why is this turned on?
k_stack_alloc_init() uses z_thread_malloc() which does not pull RAM from here, it uses the thread resource pool assigned in test_main

ajaykish replied:
@andrewboie  We want to validate k_stack_alloc_init()  api in kernel mode. Thread resource pool assignment is happening  through k_thread_resource_pool_assign() in test_main, when user space is enabled.
We are adding this test to improve code coverage on native posix platform, where user space is not enabled. That is the reason, we are using k_thread_system_pool_assign() and allocating memory from this heap memory pool.

andrewboie commented on tests/kernel/stack/stack_api/prj.conf:
```diff
@@ -2,3 +2,5 @@ CONFIG_ZTEST=y
 CONFIG_IRQ_OFFLOAD=y
 CONFIG_USERSPACE=y
 CONFIG_DYNAMIC_OBJECTS=y
+CONFIG_OBJECT_TRACING=y
```
what does CONFIG_OBJECT_TRACING have to do with this test?

nashif replied:
@ajaykish if you are doing this to increase coverage, then you should do this in a separate patch and add this to some test project that validates object_tracing, not here.

tyrantlizard77 reviewed (changes requested):
are you aware that this is already being tested?

See:

kernel/stack/stack_api/src/test_stack_contexts.c
114-
115-	zassert_not_null(stack, "couldn't allocate stack object");
116:	zassert_false(k_stack_alloc_init(stack, STACK_LEN),
117-		      "stack init failed");
118-

kernel/stack/stack_api/src/test_stack_fail.c
48-
49-	zassert_not_null(alloc_stack, "couldn't allocate stack object");
50:	zassert_false(k_stack_alloc_init(alloc_stack, STACK_LEN),
51-		      "stack init failed");
52-

ajaykish committed:
tests: kernel: Add test to validate k_stack_alloc_init, k_stack_cleanup

This test is intended to validate k_stack_cleanup() and
k_stack_alloc_init(). Passed stack pointer in k_stack_alloc_init()
is statically defined.

Signed-off-by: Ajay Kishore <ajay.kishore@intel.com>

andrewboie commented on tests/kernel/stack/stack_api/src/test_stack_contexts.c:
```diff
@@ -133,6 +133,40 @@ void test_stack_thread2isr(void)
+	int ret;
+
+#ifndef CONFIG_USERSPACE
+	k_thread_system_pool_assign(k_current_get());
```
You shouldn't need to do this.
Just remove the ifdef around the call in test_main to the k_thread_resource_pool_assign. It works without userspace unless something is broken.
